### PR TITLE
[VideoPlayer] Load program from stream property without using streaminfo

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

For PVR Streams that contain multiple streams (MPTS) the demuxer factory by default favours fast switching (not parsing streaminfo) to improve channel switch times even if the correct PID (program number) for the stream is supplied. The original PR tried to fix this by forcing reading streaminfo which was not the correct approach.

This change completely avoids reading streaminfo for TS (except for duration) and will correctly use the `program` stream property if supplied for both live and recorded streams.

Original PR which is superseded by this one:  https://github.com/xbmc/xbmc/pull/16244

Thanks for @FernetMenta for providing a solution that fixes the root cause. 

If accepted can be backported to Leia.

## Motivation and Context

In the pvr.vuplus addon certain TV providers use MPTS for their live streams. Therefore users with certain providers in some countries cannot view their channels. Example: Nos in portugal.

closes these issues:

- kodi-pvr/pvr.vuplus#195
- kodi-pvr/pvr.vuplus#208
- kodi-pvr/pvr.vuplus#219

Context:

For Enigma2 the STB simply passes the RAW TS stream. In the case of a recording the RAW stream is recorded to disk.

For the majority of providers only a single program (PID) is contained in the stream. For certain providers multiple program's are passed in the PMT but the program data packets are dropped. In this case as kodi does not parse streaminfo the correct stream is not selected.

Gist containing a log of this occurring: https://gist.github.com/phunkyfish/c6e267e57a703d0e44c4e97432415c02

In this log kodi uses Program 888 where in fact the correct program is 880.

Here is the ffmpeg output where for a bad streams where kodi would choose the wrong program. Clearly many programs are in the program table.

https://gist.github.com/phunkyfish/b1aa31816bea473168d4062796ec626c

The correct program should be 880.

Here is an example of a good stream's ffmpeg output where the only program present is 101:

https://gist.github.com/phunkyfish/1dfbf51b37f202ab75cbd4a7b28d40bb

## How Has This Been Tested?

Tested on Windows 10 and Mac OS Mojave using the pvr.vuplus addon. Tested on both Matrix and Leia.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by phunkyfish** [comment]
@hudokkow this isn't PVR specific, it would apply to any TS stream.

**Comment by Rechi** [comment]
Please either reformat the whole file in a separate commit (before the actual code changes) or only reformat the lines which contain real changes.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 1]
```diff
@@ -259,12 +260,27 @@ bool CDVDDemuxFFmpeg::Open(std::shared_ptr<CDVDInputStream> pInput, bool streami
...
+      {
+        CLog::Log(LOGDEBUG, "Error, could not open file %s", CURL::GetRedacted(strFile).c_str());
+        Dispose();
+        av_dict_free(&options);
+        return false;
+      }
       av_dict_free(&options);
-      return false;
+      m_pFormatContext->flags &= ~AVFMT_FLAG_PRIV_OPT;
+      AVDictionary *options = GetFFMpegOptionsFromInput();
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: left align `*`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 2]
```diff
@@ -1126,21 +1161,27 @@ bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double *startpts)
   int64_t seek_pts = (int64_t)time * (AV_TIME_BASE / 1000);
   bool ismp3 = m_pFormatContext->iformat && (strcmp(m_pFormatContext->iformat->name, "mp3") == 0);
 
-  if (!m_streaminfo)
-    seek_pts += m_startTime * AV_TIME_BASE;
+  if (m_checkvideo)
+  {
+    AVStream *st = m_pFormatContext->streams[m_seekStream];
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: left align `*`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 3]
```diff
@@ -1126,21 +1161,27 @@ bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double *startpts)
...
+    ret = av_seek_frame(m_pFormatContext, m_seekStream, seek_pts, backwards ? AVSEEK_FLAG_BACKWARD : 0);
 
     if (ret < 0)
     {
       int64_t starttime = m_pFormatContext->start_time;
-      if (!m_streaminfo)
-        starttime = m_startTime * AV_TIME_BASE;
+      if (m_checkvideo)
+      {
+        AVStream *st = m_pFormatContext->streams[m_seekStream];
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: left align `*`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 4]
```diff
@@ -560,14 +579,27 @@ bool CDVDDemuxFFmpeg::Open(std::shared_ptr<CDVDInputStream> pInput, bool streami
...
+  m_seekStream = -1;
+
+  if (m_checkvideo && m_streaminfo)
+  {
+    int64_t duration = m_pFormatContext->duration;
+    std::shared_ptr<CDVDInputStream> pInputStream = m_pInput;
+    Dispose();
+    bool ret = Open(pInputStream, false);
+    if (!ret)
+      return ret;
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: Eliminate the  "ret" variable.

if (!Open()....)
  return false;

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 5]
```diff
@@ -642,50 +673,50 @@ void CDVDDemuxFFmpeg::Abort()
...
+  else if (m_speed == DVD_PLAYSPEED_PAUSE && iSpeed != DVD_PLAYSPEED_PAUSE)
   {
     m_pInput->Pause(m_currentPts);
     av_read_play(m_pFormatContext);
   }
   m_speed = iSpeed;
 
   AVDiscard discard = AVDISCARD_NONE;
-  if(m_speed > 4*DVD_PLAYSPEED_NORMAL)
+  if (m_speed > 4*DVD_PLAYSPEED_NORMAL)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: space missing before and after `*`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 6]
```diff
@@ -642,50 +673,50 @@ void CDVDDemuxFFmpeg::Abort()
...
     discard = AVDISCARD_NONKEY;
-  else if(m_speed > 2*DVD_PLAYSPEED_NORMAL)
+  else if (m_speed > 2*DVD_PLAYSPEED_NORMAL)
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: space missing before and after `*`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 7]
```diff
@@ -1637,7 +1679,7 @@ CDemuxStream* CDVDDemuxFFmpeg::AddStream(int streamIdx)
     {
       stream->dvdNavId = pStream->id;
 
-      auto it = std::find_if(m_streams.begin(), m_streams.end(),
+      auto it = std::find_if (m_streams.begin(), m_streams.end(),
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: no space after `std::find_if`

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 8]
```diff
@@ -1761,14 +1803,14 @@ void CDVDDemuxFFmpeg::GetChapterName(std::string& strChapterName, int chapterIdx
...
+  if (ich)
     ich->GetChapterName(strChapterName, chapterIdx);
   else
   {
-    if(chapterIdx <= 0)
+    if (chapterIdx <= 0)
       return;
 
-    AVDictionaryEntry *titleTag = av_dict_get(m_pFormatContext->chapters[chapterIdx-1]->metadata,
+    AVDictionaryEntry* titleTag = av_dict_get(m_pFormatContext->chapters[chapterIdx-1]->metadata,
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: space before and after `-` is missing

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 9]
```diff
@@ -1810,20 +1852,20 @@ bool CDVDDemuxFFmpeg::SeekChapter(int chapter, double* startpts)
...
-  if(m_pFormatContext == NULL)
+  if (m_pFormatContext == NULL)
     return false;
 
-  if(chapter < 1 || chapter > (int)m_pFormatContext->nb_chapters)
+  if (chapter < 1 || chapter > (int)m_pFormatContext->nb_chapters)
     return false;
 
-  AVChapter *ch = m_pFormatContext->chapters[chapter-1];
+  AVChapter* ch = m_pFormatContext->chapters[chapter-1];
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: space before and after `-` is missing

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h [hunk 10]
```diff
@@ -116,17 +116,17 @@ class CDVDDemuxFFmpeg : public CDVDDemux
...
+  AVDictionary* GetFFMpegOptionsFromInput();
   double ConvertTimestamp(int64_t pts, int den, int num);
   void UpdateCurrentPTS();
   bool IsProgramChange();
   unsigned int HLSSelectProgram();
 
-  std::string GetStereoModeFromMetadata(AVDictionary *pMetadata);
-  std::string ConvertCodecToInternalStereoMode(const std::string &mode, const StereoModeConversionMap *conversionMap);
+  std::string GetStereoModeFromMetadata(AVDictionary* pMetadata);
+  std::string ConvertCodecToInternalStereoMode(const std::string &mode, const StereoModeConversionMap* conversionMap);
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h]: left align `&`

**Comment by hudokkow** [comment]
@Rechi, @ksooo, ready to go in?

**Comment by Rechi** [comment]
Jenkins tells that code style is correct now.

**Comment by phunkyfish** [comment]
@FernetMenta, would anyone else need to review this?

**Comment by fritsch** [comment]
When @FernetMenta said it's okay - it is okay :-). @ksooo you are also a reviewer here, okay to push?

**Comment by hudokkow** [comment]
I will merge this tonight if no one objects or beats me to it.

**Comment by asavah** [comment]
This PR causes massive issues with udp multicast mpegts streams.
Log (bad) : https://paste.kodi.tv/zugureleje.kodi
In this log I just open my addon (private, useful only for my ISP which provides iptv via multicast streams) and start a tv channel, nothing is played, Kodi is killed by OOM, I only press play once ... 

Log (good) same as above but with https://github.com/xbmc/xbmc/commit/e68bc9f22a576bc143014a7c3b46108a0388402f reverted
https://paste.kodi.tv/sekurexuqo.kodi

**Comment by phunkyfish** [comment]
I tested this streaming both unicast and multicast UDP and RTP streams via VLC without issue.

Maybe you can open an issue on this, and we can look at it from there.

**Comment by MilhouseVH** [comment]
I've been informed via [the forum](https://forum.kodi.tv/showthread.php?tid=343068&pid=2867577#pid2867577) of a problem since introducing this PR in RPi builds ([`#0707`](http://forum.kodi.tv/showthread.php?tid=343068&pid=2866368#pid2866368)) and x86_64 ([`#0705`](http://forum.kodi.tv/showthread.php?tid=343069&pid=2865971#pid2865971)) - streams no longer play due to a demux error.

```
2019-07-11 18:29:01.788 T:1283261312   DEBUG: GetExternalStreamDetailsFromFilename - Language = '' / Name = 'itv (External)' / Flag = '0' from /storage/.kodi/userdata/addon_data/plugin.video.itv/Subtitles/itv.srt
2019-07-11 18:29:01.788 T:1283261312  NOTICE: Creating Demuxer
2019-07-11 18:29:02.081 T:1283261312    INFO: ffmpeg[0x3568e08X]: [hls,applehttp] Opening 'https://itvpnpdotcom-a.akamaihd.net/1-8694-8528-001/18/1/VAR012/1-8694-8528-001_18_1_VAR012.ism/1-8694-8528-001_18_1_VAR012-audio=128000-video=414677.m3u8' for reading
2019-07-11 18:29:02.196 T:1543320448   DEBUG: CWebServer[8080]: request received for /jsonrpc?JSONRPC.Ping
2019-07-11 18:29:02.217 T:1543320448   DEBUG: Previous line repeats 2 times.
2019-07-11 18:29:02.217 T:1543320448   DEBUG: CWebServer[8080]: request received for /jsonrpc?Application.GetProperties
2019-07-11 18:29:02.238 T:1543320448   DEBUG: Previous line repeats 2 times.
2019-07-11 18:29:02.238 T:1543320448   DEBUG: CWebServer[8080]: request received for /jsonrpc?Player.GetActivePlayers
2019-07-11 18:29:02.253 T:1543320448   DEBUG: Previous line repeats 2 times.
2019-07-11 18:29:02.254 T:1543320448   DEBUG: CWebServer[8080]: request received for /jsonrpc?Base
2019-07-11 18:29:02.365 T:1283261312   DEBUG: Previous line repeats 2 times.
2019-07-11 18:29:02.365 T:1283261312   DEBUG: ffmpeg[0x3568e08X]: [https] HTTP error 403 Forbidden
2019-07-11 18:29:02.369 T:1283261312   DEBUG: Error, could not open file (2) https://itvpnpdotcom-a.akamaihd.net/1-8694-8528-001/18/1/VAR012/1-8694-8528-001_18_1_VAR012.ism/.m3u8?hdnea=st=1562866125~exp=1562887725~acl=/*~data=nohubplus~hmac=37451aa53db5bb6ca545b980f163a3bd6bd7ef927bcc69b7d67c03dbef8ab57f
2019-07-11 18:29:02.369 T:1283261312   ERROR: OpenDemuxStream - Error creating demuxer
2019-07-11 18:29:02.369 T:1283261312  NOTICE: CVideoPlayer::OnExit()
2019-07-11 18:29:02.371 T:1283261312   DEBUG: Thread VideoPlayer 1283261312 terminating
2019-07-11 18:29:02.392 T:1769956224    INFO: Deleting settings information for files plugin://plugin.video.itv/?url=https%3A%2F%2Fwww.itv.com%2Fhub%2Femmerdale%2F1a8694a8528&mode=3&name=Emmerdale++-+08%2F07%2F2019+18%3A00PM&iconimage=https%3A%2F%2Fhubimages.itv.com%2Fepisode%2F1_8694_8528%3Fw%3D512%26h%3D288
2019-07-11 18:29:02.396 T:1492964224   DEBUG: OnPlayBackStopped: CApplication::OnPlayBackStopped
```

[Full log](https://gist.githubusercontent.com/Mike-KL/8954dd3898d05307827832d1fd37940c/raw/0f5e2a9fe05b42b50cd5fd329107c00f69ddc6ee/gistfile1.txt)

I can reproduce the problem by playing [this URL](https://itvpnpdotcom-a.akamaihd.net/1-8694-8528-001/18/1/VAR012/1-8694-8528-001_18_1_VAR012.ism/.m3u8?hdnea=st=1562866125~exp=1562887725~acl=/*~data=nohubplus~hmac=37451aa53db5bb6ca545b980f163a3bd6bd7ef927bcc69b7d67c03dbef8ab57f
) using JSON, no add-on required.

For example:
```
texturecache.py playw "https://itvpnpdotcom-a.akamaihd.net/1-8694-8528-001/18/1/VAR012/1-8694-8528-001_18_1_VAR012.ism/.m3u8?hdnea=st=1562866125~exp=1562887725~acl=/*~data=nohubplus~hmac=37451aa53db5bb6ca545b980f163a3bd6bd7ef927bcc69b7d67c03dbef8ab57f"
```
which makes the following JSON request:
```
{"jsonrpc": "2.0", "params": {"item": {"file": "https://itvpnpdotcom-a.akamaihd.net/1-8694-8528-001/18/1/VAR012/1-8694-8528-001_18_1_VAR012.ism/.m3u8?hdnea=st=1562866125~exp=1562887725~acl=/*~data=nohubplus~hmac=37451aa53db5bb6ca545b980f163a3bd6bd7ef927bcc69b7d67c03dbef8ab57f"}}, "method": "Player.Open", "id": "libPlayer"}
```

This stream plays normally without this PR, but fails with this PR.

Reproducible on RP/RPi2/Generic (LibreELEC), and Ubuntu 17.10.

**Comment by phunkyfish** [comment]
May need @FernetMenta’s help here

**Comment by FernetMenta** [comment]
The provided URL fails on Leia, where the patch is not applied.

**Comment by MilhouseVH** [comment]
Looks like the URL has expired (curl error 22) as it no longer works wth master prior to this PR either, when it did 4 days ago. The failing add-on is pvr.iptvsimple, but I don't use that add-on (or any pvr add-on), I'm just reporting the issue - details are in the forum. Please follow up with the forum user if a new url is required.

**Comment by Mike-KL** [comment]
@FernetMenta website from which I am streaming/watching catchup is [here](https://www.itv.com/) where I utilising an old addon itvplayer which can currently be found [here](https://github.com/catchup4kodi/catchup4kodi/tree/master/zips/plugin.video.itv)  As mentioned in my forum post catchup working very well prior to #0707

Note: Requirement to be located in UK to watch these streams

As mentioned by @milhouse if you require a URL thats currently working I can provide

**Comment by phunkyfish** [comment]
@Mike-KL will have a look at the addon and see if there is anything I can spot.

Unfortunately I can’t test the stream as I’m not UK based.

**Comment by Mike-KL** [comment]
@phunkyfish just switched from nightly #0706 to #0719 with recently added PR:16369 to Properly propagate Videoplayer playback errors.  [Here](http://ix.io/1OYj) is LibreELEC's kodi.log etc containing a failure to playback addon video stream that will work fine in #0706

If you require more information just let me know.

**Comment by phunkyfish** [comment]
Ok, this PR should show the error on screen when it happens. Do you have a screen shot of it?

**Comment by Mike-KL** [comment]
screenshot000

**Comment by Mike-KL** [comment]
kodi [debug.log](https://gist.githubusercontent.com/Mike-KL/c628d4997c3a96002b2f8e6552bea792/raw/e9215864858842f1558a96bc4e1b13103d98fae2/gistfile1.txt) with verbose logging for the Video switched on. (in case it provides any additional clues to issue)

**Comment by phunkyfish** [comment]
The error is a 403, so appears to be access based. From your log this looks like file error.

```
ffmpeg[0x2cca6d0X]: [https] HTTP error 403 Forbidden
2019-07-21 12:32:13.116 T:1372750720   DEBUG: Error, could not open file (2) https://itvpnpdotcom-a.akamaihd.net/2-4586-0022-001/18/1/VAR012/2-4586-0022-001_18_1_VAR012.ism/.m3u8?hdnea=st=1563708720~exp=1563730320~acl=/*~data=nohubplus~hmac=15f770fb16d9c386ffd53ce8838f40abf58601ed87a1c9ea0cd4c053bcbc9e9f
2019-07-21 12:32:13.116 T:1372750720   ERROR: OpenDemuxStream - Error creating demuxer
```

**Comment by phunkyfish** [comment]
I also see this with VPN manager: 

```
2019-07-21 12:32:12.308 T:1633076096   DEBUG: VPN Mgr : Starting streaming media https://itvpnpdotcom-a.akamaihd.net/2-4586-0022-001/18/1/VAR012/2-4586-0022-001_18_1_VAR012.ism/.m3u8?hdnea=st=1563708720~exp=1563730320~acl=/*~data=nohubplus~hmac=15f770fb16d9c386ffd53ce8838f40abf58601ed87a1c9ea0cd4c053bcbc9e9f
2019-07-21 12:32:12.430 T:1372750720   DEBUG: GetExternalStreamDetailsFromFilename - Language = '' / Name = 'itv (External)' / Flag = '0' from /storage/.kodi/userdata/addon_data/plugin.video.itv/Subtitles/itv.srt
2019-07-21 12:32:12.430 T:1372750720  NOTICE: Creating Demuxer
2019-07-21 12:32:12.776 T:1372750720    INFO: ffmpeg[0x2cca6d0X]: [hls,applehttp] Opening 'https://itvpnpdotcom-a.akamaihd.net/2-4586-0022-001/18/1/VAR012/2-4586-0022-001_18_1_VAR012.ism/2-4586-0022-001_18_1_VAR012-audio=128000-video=447637.m3u8' for reading
```

**Comment by Mike-KL** [comment]
Yes I noticed http error 0403 Forbidden and Yes VPN Manager is playing its part in enabling playback.

I can use catchup addon in 0706 with NO issue, but from 0707 onwards I cannot. I know for sure that if I step back to 0706 build I can use addon and stream will playback OK. (Doing this frequently atm)

**Comment by phunkyfish** [comment]
But the issue is the only error we get is that there is no access unless
this is a default error message. If it's really a 403 I for the life of me
don't know why this PR would cause that. Unless there is some auth that
needs to be carried through to subsequent calls and it's not being passed.
Really not sure.

**Comment by asavah** [comment]
IMO one of the problems with this PR that it introduces reuse of context already opened with avformat_open_input() .
AFAIK this could be causing a memory leak.
Sorry I don't have logs/proofs handy.
How I found this:
Open multicast iptv stream, stop it.
Observe with tcpdump that multicast is still incoming on the nic.
Observe with netstat that Kodi is still listening on multicast socket.

FFmpeg docs state only that 
> The stream must be closed with avformat_close_input().

Dirty fix that I've been using since this PR got in, it does solves all my multicast issues:
```
diff --git a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
index f3d560cfda..6a6ee983aa 100644
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
@@ -271,6 +271,9 @@ bool CDVDDemuxFFmpeg::Open(std::shared_ptr<CDVDInputStream> pInput, bool streami
         return false;
       }
       av_dict_free(&options);
+      avformat_close_input(&m_pFormatContext);
+      m_pFormatContext  = avformat_alloc_context();
+      m_pFormatContext->interrupt_callback = int_cb;
       m_pFormatContext->flags &= ~AVFMT_FLAG_PRIV_OPT;
       AVDictionary* options = GetFFMpegOptionsFromInput();
       av_dict_set_int(&options, "load_all_variants", 0, AV_OPT_SEARCH_CHILDREN);

```

**Comment by phunkyfish** [comment]
@Mike-KL can you try with the patch and report back on results?

**Comment by phunkyfish** [comment]
@FernetMenta is what @asavah is doing a sane way to close the stream or is there a better way to handle this?

**Comment by FernetMenta** [comment]
If you start playback of a stream and stop it, CVideoPlayer::OnExit() will destruct the demuxer, which will call CDVDDemuxFFmpeg::Dispose(), which will call avformat_close_input.

I don't see a scenario that leaves m_pFormatContext open when CDVDDemuxFFmpeg::Open gets called.

If that really happens, that root cause should be fixed. The patch certainly won't fix the scenario described by @asavah:

> Open multicast iptv stream, stop it.
Observe with tcpdump that multicast is still incoming on the nic.
Observe with netstat that Kodi is still listening on multicast socket.

**Comment by asavah** [comment]
@FernetMenta I don't know too much c/c++ and my knowledge of Kodi codebase is quite tiny, but the patch I linked DOES fix the issue I described.

> I don't see a scenario that leaves m_pFormatContext open when CDVDDemuxFFmpeg::Open gets called.

That's not the point, what you describe is ofcourse correct but:

In https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L266 you open an input with a context.

And here https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L277 you open it again without closing previously open one.

It's clearly observable in case of a multicast stream that if you don't properly close the first context and open the second one "on top" of the first one - it remains open internally somewhere in the depths of libav* stuff and continues subscribed to mcast stream. Thus switching channels 20 times you'll get Kodi subscribed to 20 mcast streams ... I've had Kodi killed by OOM in a few of my tests while trying to find the cause of this.

The point is - I think that if you do avformat_open_input() you need to close without reusing it, otherwise  it will leak resources.

**Comment by FernetMenta** [comment]
@asavah sorry, I misinterpreted your diff; wasn't looking at the line numbers. Now that you linked the location it is clear. I was thinking that your patch changes code at the beginning of open() but I was wrong.
Your patch makes perfectly sense.

**Comment by phunkyfish** [comment]
@asavah if @FernetMenta is happy do you want to PR this?

**Comment by asavah** [comment]
Will do.

**Comment by Mike-KL** [comment]
@phunkyfish I can try, but how exactly would/do I apply small DVDDemuxFFmpeg.cpp patch to an installed build of libreelec? (where is the file to modify located etc)

**Comment by MilhouseVH** [comment]
@Mike-KL I'll include #16395 in the next nightly (`#0724`), it's not something that can be included unless compiling Kodi or LibreELEC.

**Comment by phunkyfish** [comment]
Thanks @MilhouseVH!

**Comment by Mike-KL** [comment]
@MilhouseVH @phunkyfish @asavah inclusion of #16395 in 0724 has fixed my issue of streaming via mentioned [add-on](https://github.com/catchup4kodi/catchup4kodi/tree/master/zips/plugin.video.itv).  LibreELEC [Kodi.log](http://ix.io/1Poi)

**Comment by phunkyfish** [comment]
Amazing news!

@hudokkow I think this is good to go in as @FernetMenta is happy with it.

@asavah is it ok with you if I just add the addition to the existing backport PR?

**Comment by MilhouseVH** [comment]
There may still be an issue with the changes introduced by this PR, please see [forum thread](https://forum.kodi.tv/showthread.php?tid=343069&pid=2873480#pid2873480) which includes a sample `.ts` video with which to reproduce the issue:

1) play from beginning
2) seek to some point in the video, say 20 mins
3) end playback, setting a resume point
4) start playback from resume point -> instant crash

There's a crash log here (latest LibreELEC/Kodi x86_64 master, not debug-enabled): http://ix.io/1Qn3

And an Ubuntu x86_64 crash log is here (latest Kodi master): http://ix.io/1Qnc

The crash issue starts when this PR is introduced in build `#0705` - fortunately there wasn't much else introduced in that build, only https://github.com/xbmc/xbmc/compare/9dfeca909b38ccd...97fe0eba17c8fd7

**Comment by phunkyfish** [comment]
Thanks for reporting @MilhouseVH. Will take a look, should be something simple. I’m won’t be next to my laptop for a few days though will look into it as soon as I can.

**Comment by phunkyfish** [comment]
Got access to a machine ;)

So the crash happens here as m_seekstream is -1: https://github.com/xbmc/xbmc/blob/6a9e668d51b890e66ea99b631eafb78baafe4e58/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L1168-L1169

The only place m_seekstream is set is to -1 at the end of Open and also to a real value in IsVideoReady(). Tried adding a call to IsVideoReady along with the check of m_checkvideo and the crash is avoided but the video just plays from the start.

**Pull request merged**